keps/sig-network/0030-nodelocal-dns-cache.md

```diff
@@ -192,7 +192,13 @@ Having the pods query the nodelocal cache introduces a single point of failure.
 
 
 ## Graduation Criteria
-TODO
+This feature has been alpha since 1.13 release and beta since 1.15 release.
+
+Graduation criteria for GA(targeted for 1.18 release):
+- Upgrade to a newer CoreDNS version(1.16.x) in [node-cache](https://github.com/kubernetes/dns/pull/328).
+- Ensure that Kubernetes [e2e tests with NodeLocal DNSCache](https://k8s-testgrid.appspot.com/sig-network-gce#gci-gce-kube-dns-nodecache) are passing.
+- Scalability tests with NodeLocal DNSCache enabled, verifying the [HA modes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190424-NodeLocalDNS-beta-proposal.md#design-details) as well as the regular mode.
+- Have N clusters(number TBD) running in production with NodeLocal DNSCache enabled.
 
```

Comment:
I'd like to get the documentation updated and improved, so that there is a Task page.

I'd like to see docs clear enough that someone who has just qualified as CKA can follow that Task page without help, and without needing in to do research / fill in missing details.

At the moment I don't think https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/ meets that criterion.

Reply:
As this is an add-on, does it also make sense to document how to remove it? (documenting that would help encourage production use)

Reply:
Thanks for the feedback! I submitted the change to docs on how to remove the addon - https://github.com/kubernetes/website/compare/master...prameshj:patch-4

Can you help improve the tasks page with suggestions on what to add? Would adding cluster create commands/kubectl commands to check that node-local-dns pods are running and healthy... make it more useful?

Reply:
From the proposed changes:
> `KUBE_ENABLE_NODELOCAL_DNS=true kubetest --up`

What are the prerequisites for that to work? (and are those explained to the reader?)

Reply:
I can try to help with documenting this add-on (I've already opened https://github.com/kubernetes/website/pull/17944). I'd want to be clear on intent here. A few things:
1. What does it mean for this add-on to be GA? I think it means that its optional and that if you enable it, the Kubernetes project believes the add-on is finished.
2. Is https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/nodelocaldns/nodelocaldns.yaml a manifest or a template that can become a manifest? What _kind_ of template?
3. How can the reader verify that their changes worked as intended?

From the proposed changes:
> `KUBE_ENABLE_NODELOCAL_DNS=true kubetest --up`

What are the prerequisites for that to work? (and are those explained to the reader?)

Reply:
@prameshj, are you planning to make a pull request based on https://github.com/kubernetes/website/compare/master...prameshj:patch-4 ?

Reply:
For this KEP, I think it's enough to add an extra line about providing clear / comprehensible documentation aimed at cluster operators.

Reply:
> kubernetes/website@master...prameshj:patch-4

I just did.. I thought I already had, thanks for the reminder!
https://github.com/kubernetes/website/pull/18190

Reply:
> I can try to help with documenting this add-on (I've already opened [kubernetes/website#17944](https://github.com/kubernetes/website/pull/17944)). I'd want to be clear on intent here. A few things:
> 
> 1. What does it mean for this add-on to be GA? I think it means that its optional and that if you enable it, the Kubernetes project believes the add-on is finished.
> 2. Is https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/nodelocaldns/nodelocaldns.yaml a manifest or a template that can become a manifest? What _kind_ of template?
> 3. How can the reader verify that their changes worked as intended?

1. IMO, Addon being GA means it is suitable to run in production.
2. It is a template, where most of the PILLAR_ variables need to be substituted. Not sure what you are referring to, by kind of template.
3. Changes can be verified by a) checking that node-local-dns pods are running b) checking the node-local-dns pods metrics to make sure request count is going up. 
> 
> From the proposed changes:
> 
> > `KUBE_ENABLE_NODELOCAL_DNS=true kubetest --up`
> 
> What are the prerequisites for that to work? (and are those explained to the reader?)
Good point. That command assumes you have a working kubernetes repo setup to build and run tests. I can include it in the other PR.

Reply:
Added documentation to the criteria.